### DISCUSSION TITLE
JSDoc comments

### DISCUSSION THREAD

**Discussion by babakfp** [open-circle/valibot#490] [Ideas]
Hi 👋

![image](https://github.com/fabian-hiller/valibot/assets/44144724/6568eda3-9845-4ed9-9f88-7d7efda6fad8)

It would be great to get those JSDoc comments to work in these types:

```ts
export type OptionsInput = v.Input<typeof OptionsSchema>
export type OptionsOutput = v.Output<typeof OptionsSchema>
```

When I pass `OptionsInput` to a function argument the user can see those comments.

(I don't know if this is even possible or not).

**Comment by fabian-hiller** [comment]
I agree! But I am not sure either. Feel free to research it if you have the time. Maybe @ssalbdivad knows the answer.

**Reply by ssalbdivad** [reply]
I worked with @Dimava a bit on this. It is possible to do in some cases but there are limitations to the scenarios where TS will preserve the association between a key and its value.

It depends a bit on what kind of transformation is being used to infer the type, but closer to a straightforward [homomorphic mapped type](https://andreasimonecosta.dev/posts/what-the-heck-is-a-homomorphic-mapped-type/) it is, the more likely it will preserve the association.

**Reply by fabian-hiller** [reply]
Okay. Thank you @ssalbdivad! Seems like this is nothing I can focus on at the moment since we have more important task with the upcoming v1 release, but it's definitely nice to have for the future if the implementation is not too "dirty".

**Reply by ssalbdivad** [reply]
@fabian-hiller In fact it's kind of the opposite- the simpler the implementation for inference, the more likely this will work.

That said, depending on what advanced mappings you might need to do, it may not be possible; it's largely not possible in ArkType because of how keys are parsed 😞

**Reply by fabian-hiller** [reply]
Then it could work for Valibot. I may will get back to you on this in a couple of months.

**Reply by Dimava** [reply]
@fabian-hiller general idea is that the thing that points to JSDoc is `keyof` and as long as you keep it's `keyof`ness the JSDoc will persist.

```ts
// keeps JSDoc
type MyOmit1<T, K> = {[P in keyof T as P extends K ? never : P]: T[P]}
// loses JSDoc
type MyOmit1<T, K> = {[P in Exclude<keyof T, K>]: T[P]}
// loses JSDoc
type MyOmit1<T, K> = {[P in (keyof T) & string]: T[P]}
// loses JSDoc
type MyOmit1<T, K> = {[P in keyof T as `${P}`]: T[P]}
```

**Reply by Dimava** [reply]
I've failed to rename keys while keeping JSDoc iirc btw

**Reply by fabian-hiller** [reply]
Thanks for the hint!
